src/compiler/crystal/semantic/type_declaration_processor.cr

```diff
@@ -280,6 +280,12 @@ struct Crystal::TypeDeclarationProcessor
         ann.raise "can't annotate #{name} in #{owner} because it was first defined in #{supervar.owner}"
       end
     else
+      owner.all_subclasses.each do |subclass|
+        if subclass.is_a?(InstanceVarContainer) && (subclass_var = subclass.instance_vars[name]?)
+          raise TypeException.new("instance variable '#{name}' of #{subclass_var.owner} is already defined in #{owner}", subclass_var.location || type_decl.location)
```

Comment:
Now that you've detected that this instance var is in a subclass, you can check if the type matches. If so, we can **remove** the instance variable from the subclass because it will be (correctly) declared in the parent.

I think this is the right fix. The specs where the type matches should compile fine, not give an error.

That said, there's no code right now to remove a declared instance vars, but it should be as easy as doing something like `subclass.instance_vars.try &.delete(name)`.

Could you give it a try? (in both places) I'm curious about the result!

Reply:
Yeah, I tried that already, but it doesn't seem to be that easy. It gave me a weird memory error.

Anyways, I think it makes sense to merge this change as is. It's an improvement and fixes the worst part. We can still figure out whether we want to have the compiler merge inherited instance variables. Let's continue that discussion in #11673.

Reply:
Makes sense! At least preventing the error from happening is a good step forward.

I think that entire module and logic should be rewritten, but it's a huge task on its own.

Reply:
That's exactly what I did as well. Makes it even more weird. I didn't try it again.

Reply:
Nevermind, found it!